src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -303,24 +307,341 @@ impl SkipWatermarkState {
+
+                read_watermarks
+                    .vnode_watermarks
+                    .iter()
+                    .map(move |(vnode, watermarks)| {
+                        (
+                            *table_id,
+                            *vnode,
+                            read_watermarks.direction,
+                            {
+                                let watermark_serde = watermark_serde.as_ref().unwrap();
+                                let row = watermark_serde
+                                .deserialize(watermarks).unwrap_or_else(|_| {
+                                    panic!("Failed to deserialize watermark {:?} serde data_types {:?} order_types {:?}", watermarks, watermark_serde.get_data_types(), watermark_serde.get_order_types());
+                                });
+                                row[0].clone()
+                            },
+                        )
+                    })
+            })
+            .collect();
+    }
+
+    fn advance_watermark(&mut self, key: &FullKey<&[u8]>) -> bool {
+        let key_table_id = key.user_key.table_id;
+        let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+        while let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front() {
+            match (table_id, vnode).cmp(&(&key_table_id, &key_vnode)) {
+                Ordering::Less => {
+                    self.remain_watermarks.pop_front();
+                    continue;
+                }
+                Ordering::Equal => {
+                    self.last_serde = self
+                        .compaction_catalog_agent_ref
+                        .watermark_serde(table_id.table_id());
+                    let (pk_prefix_serde, watermark_col_serde, watermark_col_idx_in_pk) =
+                        match self.last_serde.as_ref() {
+                            Some(serde) => serde,
+                            None => {
+                                self.last_serde = self
+                                    .compaction_catalog_agent_ref
+                                    .watermark_serde(table_id.table_id());
+                                self.last_serde.as_ref().unwrap()
+                            }
+                        };
+
+                    let row = pk_prefix_serde
+                        .deserialize(inner_key)
+                        .unwrap_or_else(|_| {
+                            panic!("Failed to deserialize pk_prefix inner_key {:?} serde data_types {:?} order_types {:?}", inner_key, pk_prefix_serde.get_data_types(), pk_prefix_serde.get_order_types());
+                        });
+                    let watermark_col_in_pk = row.datum_at(*watermark_col_idx_in_pk);
+
+                    return direction.datum_filter_by_watermark(
+                        watermark_col_in_pk,
+                        watermark,
+                        watermark_col_serde.get_order_types()[0],
+                    );
```

Comment:
The logics here is identical to the logics in `should_delete`. How about extract it into a common function?